Move runconfig/parse.go into the runconfig/opts package

dnephin opened the pull request:
This change splits the "client-side" utils from runconfig into the `runconfig/opts` package. With this change the `api/client` code only depends on `runconfig/opts` (and nothing from `runconfig`).

There are also very few imports of the `runconfig/opts` package from daemon code:

```
$ git grep 'runconfigopts\.'
builder/dockerfile/dispatchers.go:   configEnv := runconfigopts.ConvertKVStringsToMap(...
daemon/config_unix.go:   cmd.Var(runconfigopts.NewUlimitOpt(&config.Ulimits), []string{...
daemon/daemon_unix.go:       name, alias, err := runconfigopts.ParseLink(l)
```

All of these are smaller helper functions, which can be copied over, removing the need to create a `runconfig` package or having the client depend on `runconfig` from daemon.

cc @calavera

dnephin committed:
Move ParseExec to the client where it is used.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Move the runconfig.Parse() function into the runconfig/opts package.

The parse.go file is used almost exclusively in the client. The few small
functions that are used outside of the client could easily be copied out
when the client is extracted, allowing this runconfig/opts package to
move to the client.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

dnephin committed:
Move volume.SplitN() to the one place it is used in runconfig.

Signed-off-by: Daniel Nephin <dnephin@docker.com>

cpuguy83 commented on runconfig/hostconfig.go:
```diff
@@ -2,11 +2,39 @@ package runconfig
 
+var (
+	// ErrConflictContainerNetworkAndLinks conflict between --net=container and links
+	ErrConflictContainerNetworkAndLinks = fmt.Errorf("Conflicting options: container type network can't be used with links. This would result in undefined behavior")
```
Just a minor nit, but I'd either move these to `network.go` or `errors.go` instead of `hostconfig.go`

dnephin replied:
They are mostly used in `runconfig/hostconfig_unix.go`, which is why I had them in hostconfig, but `errors.go` makes sense. I'll move them

cpuguy83 commented:
LGTM.

vdemeester commented:
@dnephin cool :+1: 

Quick question :
- `runconfig.Parse` (or in this case `runconfigopts.Parse`) is only called by two method it seems, both client side (in `run.go` and `create.go`). Why not moving it to the client side too as you done for `ParseExec`.
- `ConvertKVStringsToMap` (and his companion `readKVStrings`) are used both side, but these are *utils* function, I think it would make sence to have them in `pkg/` somewhere. wdyt ?

It doesn't have to be done in this PR though I think :stuck_out_tongue_closed_eyes:

dnephin commented:
`ParseExec` was completely isolated, it didn't call any shared helper functions, so it was easy to move.

`Parse` on the other hand, does have a few shared helper functions, and calls a bunch of things in `runconfig/opts/` (ulimits, weighteddevice, throttleddevice). I didn't want to move it into client and then have daemon code importing from the client. There are a few more `opts/` functions that I'm going to move into `runconfig/opts` today as well.

I think for `ConvertKVStringsToMap` it might be easier to just copy it, since it's only used in one other place (the builder), and it is a pretty small function. If we can find a suitable package it wouldn't hurt to move it in, but I don't know of any.

icecrime commented:
Ping @anusha-ragunathan FYI!

dnephin commented:
rebased, and I moved the test fixture files into the new `runconfig/opts/fixtures` directory as well, so they're under the same directory as the tests that use them.

calavera commented:
LGTM
